src/main/java/org/candlepin/controller/ContentAccessManager.java

```diff
@@ -804,4 +820,209 @@ public Owner syncOwnerLastContentUpdate(Owner owner) {
+     *
+     * @return
+     *  the certificate or null if unable to retrieve or create a certificate
+     */
+    @Transactional
+    public AnonymousContentAccessCertificate getCertificate(AnonymousCloudConsumer consumer) {
+        if (consumer == null) {
+            throw new IllegalArgumentException("anonymous cloud consumer is null");
+        }
+
+        try {
+            AnonymousContentAccessCertificate cert = consumer.getContentAccessCert();
+            Date now = new Date();
+            // Generate a new certificate if one does not exist or the existing certificate is expired
+            if (cert == null || cert.getSerial() == null || cert.getSerial().getExpiration().before(now)) {
+                cert = createAnonContentAccessCertificate(consumer);
+            }
+
+            if (cert == null) {
+                return null;
+            }
+
+            AnonymousContentAccessCertificate wrappedCert = new AnonymousContentAccessCertificate();
+            wrappedCert.setCert(cert.getCert() + cert.getContent());
+            wrappedCert.setCreated(cert.getCreated());
+            wrappedCert.setUpdated(cert.getUpdated());
+            wrappedCert.setId(cert.getId());
+            wrappedCert.setAnonymousCloudConsumer(cert.getAnonymousCloudConsumer());
+            wrappedCert.setKey(cert.getKey());
+            wrappedCert.setSerial(cert.getSerial());
+
+            return wrappedCert;
+        }
+        catch (Exception e) {
+            String msg = "Unexpected exception occurred while fetching certificate for anonymous" +
+                "cloud consumer: {}";
+            log.error(msg, consumer, e);
+        }
+
+        return null;
+    }
+
+    private AnonymousContentAccessCertificate createAnonContentAccessCertificate(
+        AnonymousCloudConsumer consumer) throws IOException, GeneralSecurityException {
+        log.info("Generating content access certificate for anonymous cloud consumer: {}",
+            consumer.getUuid());
+
+        List<ProductInfo> products = prodAdapter.getChildrenByProductIds(consumer.getProductIds());
+        if (products == null || products.isEmpty()) {
+            log.debug("Unable to find products for anonymous cloud consumer: {}", consumer.getUuid());
+            return null;
+        }
+
+        OffsetDateTime start = OffsetDateTime.now().minus(1, ChronoUnit.HOURS);
+        OffsetDateTime end = start.plus(2, ChronoUnit.DAYS);
+
+        CertificateSerial serial = createSerial(end);
+        KeyPair keyPair = this.pki.generateKeyPair();
+        byte[] pemEncodedKeyPair = this.pki.getPemEncoded(keyPair.getPrivate());
+
+        List<ContentInfo> contentInfo = getContentInfo(products);
+        org.candlepin.model.dto.Product container = new org.candlepin.model.dto.Product();
+        List<Content> dtoContents = new ArrayList<>();
+        dtoContents.addAll((convertContentInfoToContentDto(contentInfo)));
+        container.setContent(dtoContents);
```

Comment:
You could create dtoContents list in a single step via constructor, but why do you need a new one when `convertContentInfoToContentDto` already return a new list?

Reply:
Updated